File: Pensamiento Computacional en Práctica La Lógica Universal del Código a C++, Java, C#, JavaScript y Python/Codigo GO/algoritmos_capitulo5.go
package main

import (
	"fmt"
	"strconv"
)

// CargarProductos inicializa o "carga" una matriz de productos con datos predefinidos.
func CargarProductos() [][]string {
	productos := [][]string{
		{"001", "Arroz 1kg", "35", "10"},
		{"002", "Azúcar 1kg", "25", "10"},
		{"003", "Harina 1kg", "28", "10"},
		{"004", "Aceite 1L", "50", "10"},
		{"005", "Leche 1L", "35", "10"},
		{"006", "Huevos 12 unidades", "45", "10"},
		{"007", "Fideos 500g", "20", "10"},
		{"008", "Sal 1kg", "15", "10"},
		{"009", "Pasta de tomate 400g", "25", "10"},
		{"010", "Atún lata 170g", "35", "10"},
	}
	return productos
}

// MostrarProducto formatea los datos de un único producto en una sola cadena de texto.
func MostrarProducto(vproducto []string) string {
	codigo := RellenarEspacios(vproducto[0], 5)
	producto := RellenarEspacios(vproducto[1], 30)
	precio := RellenarEspacios(vproducto[2], 10)
	cantidad := RellenarEspacios(vproducto[3], 10)
	return codigo + producto + precio + cantidad
}

// MostrarLista genera una lista completa y formateada de todos los productos de la matriz.
func MostrarLista(vproductos [][]string) string {
	var salida string
	for _, vproducto := range vproductos {
		cadena := MostrarProducto(vproducto)
		salida += cadena + "
"
	}
	return salida
}

// ExisteProducto busca un producto en la matriz por su código.
func ExisteProducto(codigo string, vproductos [][]string) int {
	for i, vproducto := range vproductos {
		if vproducto[0] == codigo {
			return i
		}
	}
	return -1
}

// ModificarProducto gestiona la interfaz y la lógica para modificar el precio de un producto.
func ModificarProducto(vproductos [][]string) {
	info := MostrarLista(vproductos)
	codigo := Leer(info + "
Introduce el codigo del producto a modificar")

	if codigo != "" {
		posicion := ExisteProducto(codigo, vproductos)

		if posicion > -1 {
			vproducto := vproductos[posicion]
			precio := Leer("
Introduce el precio de " + MostrarProducto(vproducto) + " ")

			if precio != "" {
				if EvaluarNumerico(precio, 2) || EvaluarNumerico(precio, 1) {
					vproductos[posicion][2] = precio
					fmt.Println("Producto modificado con éxito.")
				} else {
					fmt.Println("no es un valor numerico")
				}
			} else {
				fmt.Println("dato nulo")
			}
		} else {
			fmt.Println("no existe el codigo")
		}
	} else {
		fmt.Println("dato nulo")
	}
}

// MenuProductos (actualizado) muestra el menú de opciones para la gestión de productos.
func MenuProductos(vproductos [][]string) {
	datosmenuproductos := []string{"Modificar", "Listado", "Salida"}
	var opcion = "0"
	do {
		opcion = DesplegarMenu("Opciones de Productos", datosmenuproductos)
		if opcion == "" {
			fmt.Println("opcion incorrecta")
		} else {
			op, _ := strconv.Atoi(opcion)
			switch op {
			case 1:
				ModificarProducto(vproductos)
			case 2:
				fmt.Println(MostrarLista(vproductos))
			case 3:
				fmt.Println("Salida del Sistema")
			default:
				fmt.Println("No existe esta opcion")
			}
		}
	} while opcion != "3"
}

// MenuPrincipal (reestructurado) es el bucle principal de la aplicación.
func MenuPrincipal(vproductos [][]string) {
	var opcion int
	var salida bool
	datosmenuprincipal := []string{"Productos", "Punto de Venta", "Inventario", "Ventas", "Salida"}

	for !salida {
		opcionStr := DesplegarMenu("Menu de Punto de Tienda de Abarrotes la Pequeña", datosmenuprincipal)

		if opcionStr == "" {
			fmt.Println("opcion incorrecta")
		} else {
			opcion, _ = strconv.Atoi(opcionStr)
			switch opcion {
			case 1:
				MenuProductos(vproductos)
			case 2:
				// MenuPuntoVenta("000") Llama al menú de punto de venta
			case 3:
				// MenuInventario() Llama al menú de inventario
			case 4:
				fmt.Println("Ventas")
			case 5:
				fmt.Println("Salida del Sistema")
				salida = true
			default:
				fmt.Println("No existe esta opcion")
			}
		}
	}
}

File: Pensamiento Computacional en Práctica La Lógica Universal del Código a C++, Java, C#, JavaScript y Python/Codigo RUST/algoritmos_capitulo5.rs
use std::io;

fn cargar_productos() -> [[String; 4]; 10] {
    [
        [String::from("001"), String::from("Arroz 1kg"), String::from("35"), String::from("10")],
        [String::from("002"), String::from("Azúcar 1kg"), String::from("25"), String::from("10")],
        [String::from("003"), String::from("Harina 1kg"), String::from("28"), String::from("10")],
        [String::from("004"), String::from("Aceite 1L"), String::from("50"), String::from("10")],
        [String::from("005"), String::from("Leche 1L"), String::from("35"), String::from("10")],
        [String::from("006"), String::from("Huevos 12 unidades"), String::from("45"), String::from("10")],
        [String::from("007"), String::from("Fideos 500g"), String::from("20"), String::from("10")],
        [String::from("008"), String::from("Sal 1kg"), String::from("15"), String::from("10")],
        [String::from("009"), String::from("Pasta de tomate 400g"), String::from("25"), String::from("10")],
        [String::from("010"), String::from("Atún lata 170g"), String::from("35"), String::from("10")],
    ]
}

fn mostrar_producto(vproducto: &[String; 4]) -> String {
    let codigo = rellenar_espacios(&vproducto[0], 5);
    let producto = rellenar_espacios(&vproducto[1], 30);
    let precio = rellenar_espacios(&vproducto[2], 10);
    let cantidad = rellenar_espacios(&vproducto[3], 10);
    format!("{}{}{}{}", codigo, producto, precio, cantidad)
}

fn mostrar_lista(vproductos: &[[String; 4]; 10]) -> String {
    let mut salida = String::new();
    for vproducto in vproductos.iter() {
        let cadena = mostrar_producto(vproducto);
        salida.push_str(&cadena);
        salida.push('\n');
    }
    salida
}

fn existe_producto(codigo: &str, vproductos: &[[String; 4]; 10]) -> i32 {
    for (i, vproducto) in vproductos.iter().enumerate() {
        if vproducto[0] == codigo {
            return i as i32;
        }
    }
    -1
}

fn modificar_producto(vproductos: &mut [[String; 4]; 10]) {
    let info = mostrar_lista(vproductos);
    let codigo = leer(&format!("{}\nIntroduce el codigo del producto a modificar", info));

    if !codigo.is_empty() {
        let posicion = existe_producto(&codigo, vproductos);

        if posicion > -1 {
            let vproducto = &vproductos[posicion as usize];
            let precio = leer(&format!("\nIntroduce el precio de {} ", mostrar_producto(vproducto)));

            if !precio.is_empty() {
                if evaluar_numerico(&precio, 2) || evaluar_numerico(&precio, 1) {
                    vproductos[posicion as usize][2] = precio;
                    println!("Producto modificado con éxito.");
                } else {
                    println!("no es un valor numerico");
                }
            } else {
                println!("dato nulo");
            }
        } else {
            println!("no existe el codigo");
        }
    } else {
        println!("dato nulo");
    }
}

fn menu_productos(vproductos: &mut [[String; 4]; 10]) {
    let datosmenuproductos = vec!["Modificar", "Listado", "Salida"];
    let mut opcion = "0".to_string();
    while opcion != "3" {
        opcion = desplegar_menu("Opciones de Productos", &datosmenuproductos);
        if opcion.is_empty() {
            println!("opcion incorrecta");
        } else {
            let op: i32 = opcion.parse().unwrap_or(0);
            match op {
                1 => modificar_producto(vproductos),
                2 => println!("{}", mostrar_lista(vproductos)),
                3 => println!("Salida del Sistema"),
                _ => println!("No existe esta opcion"),
            }
        }
    }
}

fn menu_principal(vproductos: &mut [[String; 4]; 10]) {
    let mut opcion: i32;
    let mut salida = false;
    let datosmenuprincipal = vec!["Productos", "Punto de Venta", "Inventario", "Ventas", "Salida"];

    while !salida {
        let opcion_str = desplegar_menu("Menu de Punto de Tienda de Abarrotes la Pequeña", &datosmenuprincipal);

        if opcion_str.is_empty() {
            println!("opcion incorrecta");
        } else {
            opcion = opcion_str.parse().unwrap_or(0);
            match opcion {
                1 => menu_productos(vproductos),
                2 => {},
                3 => {},
                4 => println!("Ventas"),
                5 => {
                    println!("Salida del Sistema");
                    salida = true;
                }
                _ => println!("No existe esta opcion"),
            }
        }
    }
}
